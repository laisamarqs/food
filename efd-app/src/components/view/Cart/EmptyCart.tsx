const EmptyCart = () => (
    <p className="empty-text" style={{ color: '#ffff' }} data-testid="paragraph">
      O carrinho est√° vazio, adicione ao menos um item para continuar com a
      compra.
    </p>
  )
  
  export default EmptyCart