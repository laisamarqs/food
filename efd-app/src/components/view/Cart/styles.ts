import styled from 'styled-components'
import { ButtonAction, colors } from '../../../styles'

export const ButtonTrash = styled(ButtonAction)`
  width: 48px;
  height: 48px;
  object-fit: cover;
  bottom: 0;
  right: 0;

  img {
    width: 24px;
    height: 24px;
  }
`

export const ItemList = styled.li`
  background-color: ${colors.primaryLight};
  display: flex;
  padding: 0.5rem;
  position: relative;
  margin-top: 1rem;

  .food-image {
    width: 5rem;
    height: 5rem;
    object-fit: cover;
  }
`
export const Infos = styled.div`
  color: ${colors.primaryDark};
  margin-left: 0.5rem;

  h4 {
    text-align: center;
    font-size: 1.125rem;
    font-weight: 900;
    padding-bottom: 1rem;
  }

  span {
    font-size: 0.875rem;
    font-weight: 400;
    line-height: 1.375rem;
  }
`

export const TotalValue = styled.div`
  color: ${colors.primaryLight};
  text-align: center;
  font-size: 0.875rem;
  font-weight: 700;
  padding-top: 2.5rem;
  padding-bottom: 1rem;
  display: flex;
  justify-content: space-between;
`