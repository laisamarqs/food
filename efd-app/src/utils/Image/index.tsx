import { removeLoading } from '..'

export type ImageProps = {
  alt: string | undefined
  src: string | undefined
  width?: number
  height?: number
}

const Image = ({
  src,
  alt,
  width = 100,
  height = 100,
  ...props
}: ImageProps) => (
  <img
    loading="lazy"
    src={src}
    alt={alt}
    onLoad={removeLoading}
    className="loading"
    width={width}
    height={height}
    {...props}
  />
)

export default Image