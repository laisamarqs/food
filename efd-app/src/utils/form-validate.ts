import { FormikProps } from 'formik'

type Field =
  | 'address'
  | 'cardCode'
  | 'cardName'
  | 'cardNumber'
  | 'city'
  | 'expiresMonth'
  | 'expiresYear'
  | 'name'
  | 'number'
  | 'zipCode'

  const getFieldStatus = (field: Field, form: FormikProps<FormValues>) => {
    const isTouched = field in form.touched
    const isInvalid = field in form.errors
  
    return isTouched && isInvalid
  }
  
  export const getErrorMessage = (
    field: Field,
    form: FormikProps<FormValues>
  ): string => {
    const hasError = getFieldStatus(field, form)
  
    const errorMessage = form.errors[field]
  
    if (hasError && errorMessage) return errorMessage
    return ''
  }