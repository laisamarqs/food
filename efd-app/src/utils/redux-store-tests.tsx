import { PreloadedState } from '@reduxjs/toolkit'
import { RenderOptions, render } from '@testing-library/react'
import { PropsWithChildren, ReactElement } from 'react'
import { Provider } from 'react-redux'

import { RootState, AppStore, configStore } from '../store'
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProvider(
  element: ReactElement,
  {
    preloadedState = {},
    store = configStore(preloadedState),
    ...moreOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }

  return {
    store,
    ...render(element, {
      wrapper: Wrapper,
      ...moreOptions
    })
  }
}