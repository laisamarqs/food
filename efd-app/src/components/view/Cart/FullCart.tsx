import { useDispatch, useSelector } from 'react-redux'

import Button from '../../ui/Button'

import * as Styles from './styles'
import trashIcon from '../../../assets/images/trash-icon.svg'
import { calculeTotalPrice, formatPrice } from '../../../utils'

import { RootState } from '../../../store'
import { removeOfCart } from '../../../store/reducers/cart'
import { changeComponent } from '../../../store/reducers/sideBar'

const FullCart = () => {
  const { itemsCart } = useSelector((state: RootState) => state.cart)
  const dispatch = useDispatch()

  return (
    <>
      <ul>
        {itemsCart.map((item) => (
          <Styles.ItemList key={item.id}>
            <img src={item.foto} alt={item.nome} className="food-image" />
            <Styles.Infos>
              <h4>{item.nome}</h4>
              <span>{formatPrice(item.preco)}</span>
            </Styles.Infos>
            <Styles.ButtonTrash
              onClick={() => dispatch(removeOfCart(item.id))}
              title="Remover o item do carrinho"
            >
              <img src={trashIcon} alt="icone de lixeira da cor rosa" />
            </Styles.ButtonTrash>
          </Styles.ItemList>
        ))}
      </ul>

      <Styles.TotalValue>
        <span>Valor Total</span>
        <span>{formatPrice(calculeTotalPrice(itemsCart))}</span>
      </Styles.TotalValue>

      <Button
        title="Continuar com a entrega"
        label="Continuar com a entrega"
        background="secundary"
        type="button"
        onClick={() => dispatch(changeComponent('form'))}
      />
    </>
  )
}

export default FullCart