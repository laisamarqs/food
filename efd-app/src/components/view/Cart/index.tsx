import { useSelector } from 'react-redux'

import FullCart from './FullCart'
import EmptyCart from './EmptyCart'

import { RootState } from '../../../store'

const Cart = () => {
  const { itemsCart } = useSelector((state: RootState) => state.cart)

  return <div>{itemsCart.length > 0 ? <FullCart /> : <EmptyCart />}</div>
}

export default Cart