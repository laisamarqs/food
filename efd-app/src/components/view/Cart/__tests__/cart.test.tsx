import '@testing-library/jest-dom'
import { fireEvent, screen } from '@testing-library/dom'

import Cart from '..'
import { renderWithProvider } from '../../../../utils/redux-store-tests'
import { expect } from 'vitest'

const renderComponent = () => {
  const { container, store } = renderWithProvider(<Cart />, {
    preloadedState: {
      cart: {
        itemsCart: [
          {
            foto: 'img/img.png',
            preco: 12.7,
            id: 1,
            nome: 'comida 1',
            descricao: 'descrição 1',
            porcao: '2 pessoas'
          },
          {
            foto: 'img/img.png',
            preco: 129,
            id: 2,
            nome: 'comida 2',
            descricao: 'descrição 2',
            porcao: '2 pessoas'
          }
        ]
      }
    }
  })
  return { container, store }
}

describe('<Cart />', () => {
  it('should render without itens', () => {
    const { container } = renderWithProvider(<Cart />, {
      preloadedState: { cart: { itemsCart: [] } }
    })

    expect(screen.getByTestId('paragraph')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with 2 items in cart', () => {
    const { container } = renderComponent()

    const heading1 = screen.getByRole('heading', { name: 'comida 1' })
    const heading2 = screen.getByRole('heading', { name: 'comida 2' })
    expect(heading1).toBeInTheDocument()
    expect(heading2).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should remove an item of cart when clicking on the trash button', () => {
    const { store } = renderComponent()

    const trashButton = screen.getAllByRole('button', {
      name: 'icone de lixeira da cor rosa'
    })

    fireEvent.click(trashButton[0])

    expect(store.getState().cart.itemsCart.length).toEqual(1)
  })

  it('should call a function when clicking continue with delivery', () => {
    const { store } = renderComponent()

    const keepWithDeliveryButton = screen.getByRole('button', {
      name: /Continuar com a entrega/i
    })

    fireEvent.click(keepWithDeliveryButton)

    expect(store.getState().sideBar.component).toBe('form')
  })
})