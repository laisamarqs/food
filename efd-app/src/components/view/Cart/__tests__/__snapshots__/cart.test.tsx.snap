// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Cart /> > should render with 2 items in cart 1`] = `
<div>
  <ul>
    <li
      class="sc-gsFSXq enYCKw"
    >
      <img
        alt="comida 1"
        class="food-image"
        src="img/img.png"
      />
      <div
        class="sc-kAyceB fygJfL"
      >
        <h4>
          comida 1
        </h4>
        <span>
          R$ 12,70
        </span>
      </div>
      <button
        class="sc-eqUAAy sc-iGgWBj KzdVX euZauh"
        title="Remover o item do carrinho"
      >
        <img
          alt="icone de lixeira da cor rosa"
          src="/src/assets/images/trash-icon.svg"
        />
      </button>
    </li>
    <li
      class="sc-gsFSXq enYCKw"
    >
      <img
        alt="comida 2"
        class="food-image"
        src="img/img.png"
      />
      <div
        class="sc-kAyceB fygJfL"
      >
        <h4>
          comida 2
        </h4>
        <span>
          R$ 129,00
        </span>
      </div>
      <button
        class="sc-eqUAAy sc-iGgWBj KzdVX euZauh"
        title="Remover o item do carrinho"
      >
        <img
          alt="icone de lixeira da cor rosa"
          src="/src/assets/images/trash-icon.svg"
        />
      </button>
    </li>
  </ul>
  <div
    class="sc-imWYAI iyRkcB"
  >
    <span>
      Valor Total
    </span>
    <span>
      R$ 141,70
    </span>
  </div>
  <button
    class="sc-dcJsrY bDWePn"
    title="Continuar com a entrega"
    type="button"
  >
    Continuar com a entrega
  </button>
</div>
`;

exports[`<Cart /> > should render without itens 1`] = `
<div>
  <p
    class="empty-text"
    data-testid="paragraph"
    style="color: rgba(255, 255, 255, 1);"
  >
    O carrinho está vazio, adicione ao menos um item para continuar com a compra.
  </p>
</div>
`;